File: 04-services/api-gateway/src/main/java/com/example/apigateway/AppRuntimeHints.java
package com.example.apigateway;

import org.springframework.aot.hint.RuntimeHints;
import org.springframework.aot.hint.RuntimeHintsRegistrar;

public class AppRuntimeHints implements RuntimeHintsRegistrar {

    @Override
    public void registerHints(RuntimeHints hints, ClassLoader classLoader) {
        // must be registered by Spring Cloud Gateway, bug fix
        hints.resources().registerPattern("META-INF/scripts/request_rate_limiter.lua");
    }
}

File: 04-services/delivery-service/src/main/kotlin/com/example/deliveryservice/config/AppRuntimeHints.kt
package com.example.deliveryservice.config

import liquibase.database.LiquibaseTableNamesFactory
import liquibase.parser.SqlParserFactory
import liquibase.report.ShowSummaryGeneratorFactory
import liquibase.ui.LoggerUIService
import org.springframework.aot.hint.ExecutableMode
import org.springframework.aot.hint.RuntimeHints
import org.springframework.aot.hint.RuntimeHintsRegistrar
import java.util.*

class AppRuntimeHints : RuntimeHintsRegistrar {
    override fun registerHints(hints: RuntimeHints, classLoader: ClassLoader?) {
        hints.resources().registerPattern("liquibase/*")
        hints.reflection()
            .registerType(LoggerUIService::class.java) { type ->
                type.withConstructor(Collections.emptyList(), ExecutableMode.INVOKE)
            }
            .registerType(LiquibaseTableNamesFactory::class.java) { type ->
                type.withConstructor(Collections.emptyList(), ExecutableMode.INVOKE)
            }
            .registerType(ShowSummaryGeneratorFactory::class.java) { type ->
                type.withConstructor(Collections.emptyList(), ExecutableMode.INVOKE)
            }
            .registerType(SqlParserFactory::class.java) { type ->
                type.withConstructor(Collections.emptyList(), ExecutableMode.INVOKE)
            }
    }
}